webui/react/src/hooks/useSettings.ts

```diff
@@ -0,0 +1,160 @@
+import { useHistory, useLocation } from 'react-router-dom';
+
+import { Primitive } from 'types';
+import { clone } from 'utils/data';
+
+import usePrevious from './usePrevious';
+import useStorage from './useStorage';
+
+export enum BaseType {
+  Boolean = 'Boolean',
+  Float = 'Float',
+  Integer = 'Integer',
+  String = 'String',
+}
+
+type GenericSettingsType = Primitive | Primitive[] | undefined;
+
+interface SettingsConfigProp {
+  defaultValue: GenericSettingsType;
+  isFilter?: boolean;
+  key: string;
+  storageKey?: string;
+  type: {
+    baseType: BaseType;
+    isArray?: boolean;
+  };
+}
+
+export interface SettingsConfig {
+  settings: SettingsConfigProp[];
+  storagePath: string;
+}
+
+type GenericSettings = Record<string, GenericSettingsType>;
+
+interface SettingsHook<T> {
+  filterCount: number;
+  resetFilters: () => void;
+  settings: T;
+  updateSettings: (newSettings: Partial<T>) => void;
+}
+
+const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
+  if (param === null) return undefined;
+  if (type === BaseType.Boolean) return param === 'true';
+  if (type === BaseType.Float || type === BaseType.Integer) {
+    const value = type === BaseType.Float ? parseFloat(param) : parseInt(param);
+    return !isNaN(value) ? value : undefined;
+  }
+  if (type === BaseType.String) return param;
+  return undefined;
+};
+
+const queryToSettings = <T>(
+  query: string,
+  config: SettingsConfig,
+  defaultSettings?: GenericSettings,
+): T => {
+  const params = queryString.parse(query);
+  return config.settings.reduce((acc, prop) => {
+    /*
+     * Attempt to decode the query parameter and if anything
+     * goes wrong, set it to the default value.
+     */
+    try {
+      const paramValue = params[prop.key];
+      const baseType = prop.type.baseType;
+
+      /*
+       * Convert the string-based query params to primitives.
+       * `undefined` values can happen if the query param values are invalid.
+       *   string[] => Primitive[] | undefined
+       *   string   => Primitive | undefined
+       *   null     => undefined
+       */
+      const queryValue = Array.isArray(paramValue)
+        ? paramValue
+          .map(value => queryParamToType(baseType, value))
+          .filter((value): value is Primitive => value !== undefined)
+        : queryParamToType(baseType, paramValue);
+
+      /*
+       * When expecting an array, convert valid non-array values into an array.
+       * Example - 'PULLING' => [ 'PULLING' ]
+       */
+      const normalizedValue = prop.type.isArray && queryValue != null && !Array.isArray(queryValue)
+        ? [ queryValue ] : queryValue;
+
+      acc[prop.key] = normalizedValue;
+    } catch (e) {
+      if (defaultSettings && defaultSettings[prop.key] !== undefined) {
+        acc[prop.key] = defaultSettings[prop.key];
+      }
+    }
+    return acc;
+  }, {} as GenericSettings) as unknown as T;
```

Comment:
Open to ideas on how to avoid this `X as unknown as Y` type casting. This goal in general is to cast it to something that's either more specific or something that's more general depending on the use case. The one here specifically is going the direction of more specific. Where we process the query string into a `GenericSettings` but we know for a fact, given the config, it is actually a specific type `T`.